### PR TITLE
Global Authentication (built in Hapi)

### PR DESCRIPTION
**Pull request by kenkouot**
Alrighty, so this is foundational work for the new global login page that the social team is doing. This PR provides a global login page that will validate your credentials with Helios and redirect to a route `/auth` implemented in your application. The `/auth` route allows your respective app to handle the credentials however you want (set your session handling, etc..,)

## Sample Flow
For instance, take this (hypothetical) flow in Mercury:

1. Start with an unauthenticated user
2. User attempts to access `//muppet.wikia.com/wiki/Elmo/edit` on Mercury *(VCL logic puts this user in Mercury test group, so requests get routed to Mercury backends)*
2. Application redirects to `//www.wikia.com/login?redirect=muppet.wikia.com/wiki/Elmo/edit` (New feature) and is served `/login` by Mercury
3. User provides credentials through log in form, hits submit
4. Mercury gets the POST, successfully grabs credentials from Helios
5. Mercury uses these credentials to set cookies
6. Mercury then redirects the user to the URI specified in the `redirect` param. MediaWiki and Mercury Ember apps are expected to understand these cookies.

## CSRF Protection
This PR also adds CSRF protection by default in our hapi routes. AJAX requests to `/login' will also need to provide a CSRF token as part of their request payload.

## Special Notes
* Helios config is stored in localSettings, which is .gitignore'd, please do not store any sensitive data in app code.

## Migration Strategy
* This PR also introduces changes to i18n usage. Please check #615 for details.

### Remaining Tasks
* [x] Add CSRF protection using [Crumb](https://github.com/hapijs/crumb)
* [x] Document new routes and expectations for referring clients
* [x] Normalize error states sent from server to client for login page
* [x] Implement "Remember Me"

* [x] Front-End for the Landing Page
* [x] Front-End for the Login Page
* [x] Validation for the Login Page
* [.] Front-End for the Signup Page
* [.] Validation for the Signup Page
* [.] Translations for the Login / Signup

### CODE REVIEW

**Comment by kenkouot** [comment]
@ArturKlajnerok @frankfarmer @garthwebb @lizlux @jsutterfield @cmiller-wikia

File: server/app.ts [hunk 1]
```diff
@@ -7,207 +7,218 @@ if (process.env.NEW_RELIC_ENABLED === 'true') {
...
+			failAction: 'log'
+		}
+	}
+});
+
+setupLogging(server);
+
+var plugins = [
+	{register: require('hapi-auth-cookie')},
+	{register: require('crumb'), options: {
```
Comment (lizlux) [inline server/app.ts]: Spacing here looks weird to me. Doesn't really follow [conventions](https://github.com/Wikia/guidelines/blob/master/JavaScript/CodingConventions.md#objects).

File: server/facets/auth/handleAuth.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
+/// <reference path='../../../typings/hapi/hapi.d.ts' />
+var rememberMeTTL = 2419200 * 1000; // 4 weeks
```
Comment (lizlux) [inline server/facets/auth/handleAuth.ts]: I believe our current "remember me" value is 6 months. Was this change intentional?

**Comment by lizlux** [comment]
@kenkouot, we should probably include the "remember me" boolean in the JSON response so clients can set their own auth TTL.

File: server/facets/auth/handleAuth.ts (hunk 2 above)
Comment (mklucsarits) [inline server/facets/auth/handleAuth.ts]: We can change it to whatever is appropriate. This was temporary until I found out the correct value.

File: server/facets/auth/handleAuth.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+/**
+ * Stores the user's authentication data, which is passed along from a login request
+ *
+ * @param {Hapi.Request} request
+ * @param {Hapi.Response} reply
+ */
+function storeAuthentication (request: Hapi.Request, reply: any): void {
```
Comment (mklucsarits) [inline server/facets/auth/handleAuth.ts]: When would ```reply``` not be a Hapi.Response object?

File: server/routes.ts [hunk 4]
```diff
@@ -12,6 +20,15 @@ var unauthenticatedRoutes: any[],
 			cache: {
 				privacy: Caching.policyString(Caching.Policy.Public),
 				expiresIn: 60000
+			},
+			auth: {
+				mode: 'try',
+				strategy: 'session'
+				},
```
Comment (mklucsarits) [inline server/routes.ts]: Indentation looks wrong here.

File: server/facets/auth/handleAuth.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
+/// <reference path='../../../typings/hapi/hapi.d.ts' />
+var rememberMeTTL = 1.57785e10 // 6 months
```
Comment (mklucsarits) [inline server/facets/auth/handleAuth.ts]: missing semicolon

File: config/localSettings.base.ts [hunk 6]
```diff
@@ -17,6 +17,13 @@ var localSettings: LocalSettings = {
 	// Targeted environment [prod|preview|verify|dev|testing]
 	environment: Utils.getEnvironment(process.env.WIKIA_ENVIRONMENT),
 	// NOTE: On your devbox, use your eth0 address in able to bind route to something accessible
```
Comment (lizlux) [inline config/localSettings.base.ts]: This comment should go below the helios config

File: server/app.ts [hunk 7]
```diff
@@ -7,207 +7,223 @@ if (process.env.NEW_RELIC_ENABLED === 'true') {
...
-				});
+var server = new Hapi.Server();
+
+server.connection({
+	host: localSettings.host,
+	port: localSettings.port,
+	routes: {
+		state: {
+			// We currently don't use any cookies on server side
+			// Uncomment this setting if you change the one above as we don't want to fail on invalid cookies
```
Comment (lizlux) [inline server/app.ts]: Should these two comments be updated?

File: server/facets/auth/login.ts [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+export function get (request: Hapi.Request, reply: any): void {
+	if (request.auth.isAuthenticated) {
+		return reply.redirect(request.query.redirect || '/');
+	}
+
+	return reply.view('login', null, {
+		layout: 'wikia-static'
+	});
+};
```
Comment (lizlux) [inline server/facets/auth/login.ts]: don't need semicolon here

File: server/facets/auth/login.ts [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+		authRedirect = '/auth?' + qs.stringify(authParams);
+
+		if (isAJAX) {
+			return reply({authRedirect: authRedirect});
+		}
+
+		return reply.redirect(authRedirect);
+	});
+};
```
Comment (lizlux) [inline server/facets/auth/login.ts]: don't need semicolon here

File: server/facets/auth/login.ts [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+		layout: 'wikia-static'
+	});
+};
+
+export function post (request: Hapi.Request, reply: any): void {
+	var credentials: any = request.payload,
+		authParams: AuthParams,
+		requestedWithHeader: string = request.headers['x-requested-with'],
+		isAJAX: boolean = requestedWithHeader && !!requestedWithHeader.match('XMLHttpRequest'),
+		error: any = {},
```
Comment (lizlux) [inline server/facets/auth/login.ts]: Looks like this variable isn't being used.

File: server/routes.ts [hunk 11]
```diff
@@ -75,8 +137,21 @@ articlePagePaths.forEach((path) => {
...
+		path: '/test',
+		// TODO: This is just an example, remove this handler logic from the routes file later
+		handler (request: Hapi.Request, reply: any) {
+			reply(request.auth);
+		}
+	}
+];
+
 unauthenticatedRoutes = unauthenticatedRoutes.map((route) => {
 	return Hoek.applyToDefaults(unauthenticatedRouteConfig, route);
```
Comment (lizlux) [inline server/routes.ts]: Are we trying to name these local variables after the module that gets imported? So for example, this `Hoek` should actually be defined as `hoek`?

File: server/facets/operations/generateCSRFView.ts [hunk 12]
```diff
@@ -0,0 +1,6 @@
+/// <reference path='../../../typings/hapi/hapi.d.ts' />
+function generateCSRFView (request: Hapi.Request, reply: any): void {
+	reply.view('breadcrumb', null, {layout: 'empty'});
```
Comment (lizlux) [inline server/facets/operations/generateCSRFView.ts]: Do we need to add a `crumb` value here for the template? Or is that still a WIP?

Comment (lizlux) [inline server/facets/operations/generateCSRFView.ts]: Saw in the [docs](https://www.npmjs.com/package/crumb) that `crumb` is already exported to the view. Magic!

File: front/styles/auth/base/_body.scss [hunk 13]
```diff
@@ -0,0 +1,9 @@
+body {
+	background-color: $color-blue-dark;
+	display: flex;
+	flex-direction: column;
+	height: 100vh;
```
Comment (bkoval) [inline front/styles/auth/base/_body.scss]: I would consider changing these to 100%, as Android Default Browser started to support these from version 4.4 and I think a big amount of our users has JellyBean / Ice Cream Sandwich on their deviced. What do yout hink, @lizlux ? http://caniuse.com/#search=viewport%20units

File: server/routes.ts (hunk 11 above)
Comment (kenkouot) [inline server/routes.ts]: No, I don't think that is an actual convention. I name them this because that's the Hoek/Hapi convention. They just follow the convention of classes, being considered static utility classes. Npm modules are all lower dash cased, so it would be totally arbitary to follow their casing.

File: server/facets/auth/handleAuth.ts [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+	}
+
+	request.auth.session.set({
+		user_id       : request.query.user_id,
+		access_token  : request.query.access_token,
+		refresh_token : request.query.refresh_token
+	});
+
+	if (request.query.remember) {
+		request.auth.session.ttl(rememberMeTTL);
```
Comment (kvas-damian) [inline server/facets/auth/handleAuth.ts]: What was the reason to implement "remember me" by increasing TTL instead of creating separate long-lived cookie with token?

According to document created by @cmiller-wikia :
> A long-lived session token is associated with that user, **in a different cookie to the normal login session**. 
> https://one.wikia-inc.com/wiki/Engineering/Login_Sessions_and_Services#Baseline_Security_Assumptions

**Comment by kenkouot** [comment]
Testing this today and am confident that we can merge this, safe from regressions affecting the main Ember app.

Going to merge this in the morning if nothing huge is outstanding.

File: server/routes.ts [hunk 15]
```diff
@@ -80,8 +141,21 @@ articlePagePaths.forEach((path) => {
 	});
 });
 
+authenticatedRoutes = [
+	{
+		method: 'GET',
+		path: '/test',
+		// TODO: This is just an example, remove this handler logic from the routes file later
```
Comment (rogatty) [inline server/routes.ts]: Do we want to merge it to dev?

File: server/facets/auth/login.ts [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+		redirect = request.query.redirect || '/';
+
+		request.auth.session.set({
+			'user_id'       : response.user_id,
+			'access_token'  : response.access_token,
+			'refresh_token' : response.refresh_token
+		});
+
+		if (credentials.remember) {
+			request.auth.session.ttl(rememberMeTTL);
```
Comment (kvas-damian) [inline server/facets/auth/login.ts]: [this comment](https://github.com/Wikia/mercury/pull/592#discussion-diff-27406689) is still not addressed

File: server/views/_layouts/wikia-static.hbs [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+		<!-- endbuild -->
+	</head>
+	<body class="auth {{bodyClasses}}">
+		{{#if exitTo}}
+		<a class="close" href="{{exitTo}}"></a>
+		{{/if}}
+
+		{{#unless hideHeader }}
+		{{!-- TODO: use i18n for messages once it's set up (SOC-526) --}}
+		<header class="auth-header splash splash-top">Welcome Back!</header>
```
Comment (kvas-damian) [inline server/views/_layouts/wikia-static.hbs]: Looks like there are many not translated messages that are going to be merged to dev.

File: server/facets/auth/login.ts (hunk 16 above)
Comment (kenkouot) [inline server/facets/auth/login.ts]: The TTL cookie will be a separate cookie. I'd consider it still a WIP and rest assured, I've been working directly with @cmiller-wikia and @gbenson-wikia to make sure that it gets executed with their blessing :dancer:

File: server/views/_layouts/wikia-static.hbs (hunk 17 above)
Comment (lizlux) [inline server/views/_layouts/wikia-static.hbs]: This is a known issue and there's a ticket for it https://wikia-inc.atlassian.net/browse/SOC-526. We had to hard code some English messages during development while server-side i18n was still being set up.

File: server/routes.ts (hunk 15 above)
Comment (lizlux) [inline server/routes.ts]: Good catch @rogatty, this should be removed or commented out before we merge .

**Comment by lizlux** [comment]
Just to be clear, we'll merge this PR to dev but the new auth pages will not be user-facing until the VCL logic is in place to handle these routes.

File: front/styles/auth/base/_input.scss [hunk 18]
```diff
@@ -0,0 +1,12 @@
+input[type='text'],
+input[type='text']:focus,
+input[type='password'],
+input[type='password']:focus {
```
Comment (mklucsarits) [inline front/styles/auth/base/_input.scss]: Why aren't these handled by the style guide?

File: front/styles/auth/base/_lists.scss [hunk 19]
```diff
@@ -0,0 +1,8 @@
+ul, li {
```
Comment (mklucsarits) [inline front/styles/auth/base/_lists.scss]: I thought the style guide reset these already.

File: server/views/auth-landing-page.hbs [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+                Connect with Facebook
+            </a>
+        </li>
+        <li>
+            <a class="button signup-provider-email">
+                Sign up with email
+            </a>
+        </li>
+    </ul>
+        <p class="sign-in-message">
```
Comment (mklucsarits) [inline server/views/auth-landing-page.hbs]: If you use a `small` tag instead of `p` here, the font will be sized by the style guide at 0.75 rem, and we won't need to include a hard-coded font-size for this element.

One of the goals of the style guide is to have a font scale and stick to it. Ideally, if the style guide CSS is being included, there should be no need to specify a `font-size` declaration for anything. And if you come upon a font size that isn't part of our font scale in the style guide, go back to the designer and tell them, because it shouldn't be used.

File: front/svg/images/wikia-logo-320.svg [hunk 21]
```diff
@@ -0,0 +1 @@
...up"/><path d="M321.2 404.3 L321.7 396.7 L313.7 396.7 L313.1 399.6 C311.3 397.6 309 396 304.9 396 C297.6 396 293.6 400.7 293.6 409.7 C293.6 418.7 297.6 423.4 304.9 423.4 C309 423.4 311.3 421.8 313.1 419.8 L313.7 422.8 L321.7 422.8 L321.2 415.1 L321.2 404.3 L321.2 404.3 ZM312.5 413 C311.3 414.5 309.6 415.5 307.3 415.5 C304.7 415.5 302.6 413.7 302.6 409.7 C302.6 405.7 304.7 404 307.3 404 C309.6 404 311.4 404.9 312.5 406.4 L312.5 413 L312.5 413 Z" sketch:type="MSShapeGroup"/><path d="M230.1 393.9 C232.9 393.9 235.2 391.6 235.2 388.9 C235.2 386.1 233 383.8 230.2 383.8 C227.4 383.8 225.2 386.1 225.2 388.9 C225.1 391.6 227.4 393.9 230.2 393.9 " sketch:type="MSShapeGroup"/><path d="M225.1 396.7 L225.1 422.8 L237.7 422.8 L237.7 417.3 L233.8 417.3 L233.8 402.2 L237.7 402.2 L237.7 396.7 L225.1 396.7 " sketch:type="MSShapeGroup"/><path d="M284.2 383.8 C281.4 383.8 279.2 386.1 279.2 388.9 C279.2 391.6 281.5 393.9 284.2 393.9 C287 393.9 289.3 391.6 289.3 388.9 C289.2 386.1 287 383.8 284.2 383.8 " sketch:type="MSShapeGroup"/><path d="M276.6 396.7 L276.6 402.2 L280.6 402.2 L280.6 417.3 L276.6 417.3 L276.6 422.8 L289.2 422.8 L289.2 396.7 L276.6 396.7 " sketch:type="MSShapeGroup"/></g></g></g></svg>
```
Comment (mklucsarits) [inline front/svg/images/wikia-logo-320.svg]: There's a lot of extraneous markup in this file from Sketch that should be removed.

File: front/styles/auth/base/_input.scss (hunk 18 above)
Comment (lizlux) [inline front/styles/auth/base/_input.scss]: These input styles actually differ from the style guide, which isn't awesome, but should be taken up with Witcher if there's concern.

File: server/views/auth-landing-page.hbs (hunk 20 above)
Comment (kenkouot) [inline server/views/auth-landing-page.hbs]: Agree with Matt here. Also, we might want to introduce a .small class (I think there already might be one...)

File: front/svg/images/wikia-logo-320.svg (hunk 21 above)
Comment (kenkouot) [inline front/svg/images/wikia-logo-320.svg]: Does that get processed out with the svg optimizer?

File: server/app.ts [hunk 22]
```diff
@@ -7,207 +7,237 @@ if (process.env.NEW_RELIC_ENABLED === 'true') {
...
+	}
+	return reply.continue();
+});
+
+server.on('tail', () => {
+	counter++;
+
+	if (counter >= localSettings.maxRequestsPerChild) {
+		//This is a safety net for memory leaks
+		//It restarts child so even if it leaks we are 'safe'
```
Comment (inez) [inline server/app.ts]: True. But it also makes it harder to notice that application is leaking. I don't know what's the default value for maxRequestsPerChild but I suggest making it "infinite" and using this feature only when really needed (not by default).

File: server/views/login.hbs [hunk 23]
```diff
@@ -0,0 +1,18 @@
+<section class="auth-content">
+	<span>{{json error}}</span>
+	<form id="loginForm" action="" method="POST">
+		<div class="input-container">
+			<input type="text" name="username" id="loginUsername" required />
+			<label class="floating-label" for="loginUsername">Username</label>
+		</div>
+		<div class="input-container">
+			<input class="sign-in-password" type="password" name="password" id="loginPassword" required />
+			<label class="floating-label" for="loginPassword">Password</label>
```
Comment (inez) [inline server/views/login.hbs]: Hardcoded English strings

File: server/views/_layouts/wikia-static.hbs [hunk 24]
```diff
@@ -0,0 +1,33 @@
+<!DOCTYPE html>
+<html>
+	<head>
+		<meta charset="utf-8">
+		<meta name="viewport" content="user-scalable=no, width=device-width, initial-scale=1, maximum-scale=1, minimal-ui">
+
+		<title>{{{title}}}</title>
+		<!-- build:css /front/styles/auth/app.css -->
+		<link rel="stylesheet" href="/front/styles/auth/app.css">
```
Comment (mklucsarits) [inline server/views/_layouts/wikia-static.hbs]: It doesn't seem like CSS in the /auth directory should be coupled with the wikia-static.hbs template.

File: server/views/login.hbs [hunk 25]
```diff
@@ -0,0 +1,18 @@
+<section class="auth-content">
```
Comment (mklucsarits) [inline server/views/login.hbs]: The section element isn't supposed to be used as a generic container. I think a div would be more suitable here.

File: server/views/login.hbs (hunk 23 above)
Comment (lizlux) [inline server/views/login.hbs]: https://github.com/Wikia/mercury/pull/592/files#r27500283

File: server/app.ts [hunk 26]
```diff
@@ -7,207 +7,237 @@ if (process.env.NEW_RELIC_ENABLED === 'true') {
...
 }
 
-import Hapi = require('hapi');
-import path = require('path');
-import url = require('url');
+import Caching       = require('./lib/Caching');
+import Hapi          = require('hapi');
+import Logger        = require('./lib/Logger');
+import Utils         = require('./lib/Utils');
+import cluster       = require('cluster');
```
Comment (mklucsarits) [inline server/app.ts]: Is there any convention here for what's capitalized or not?

File: server/views/login.hbs (hunk 25 above)
Comment (bkoval) [inline server/views/login.hbs]: Agreed. We'll change it.

File: server/views/_layouts/wikia-static.hbs (hunk 24 above)
Comment (bkoval) [inline server/views/_layouts/wikia-static.hbs]: It is. I can see how the template name can be misleading, let's change it to auth-static or sth. more matching the purpose of the template.

File: server/app.ts (hunk 26 above)
Comment (kenkouot) [inline server/app.ts]: Handling this in separate ticket, but I agree the convention could be clearer.
